internal/provider/factory.go

```diff
@@ -68,76 +72,91 @@ type Factory interface {
+	providers              []string
+	delegationProviderFunc DelegationProviderFunc
 }
 
 // NewFactory returns a new provider factory with the given client and given providers enabled.
 // Will return an error if any given provider has no registered provider implementation.
-func NewFactory(c client.Client, d dynamic.Interface, p []string) (Factory, error) {
+func NewFactory(c client.Client, d dynamic.Interface, p []string, dpf DelegationProviderFunc) (Factory, error) {
 	var err error
 	registeredProviders := maps.Keys(constructors)
 	for _, provider := range p {
 		if !slices.Contains(registeredProviders, provider) {
 			err = errors.Join(err, fmt.Errorf("provider '%s' not registered", provider))
 		}
 	}
-	return &factory{Client: c, dynamicClient: d, providers: p}, err
+	return &factory{Client: c, dynamicClient: d, providers: p, delegationProviderFunc: dpf}, err
 }
 
-// ProviderFor will return a Provider interface for the given ProviderAccessor secret.
-// If the ProviderAccessor is delegating an in memory endpoint provider secret is returned with configuration appropriate for delegation of that resource.
-// If the requested ProviderAccessor is not delegating, and the secret does not exist in the resource namespace, an error will be returned.
-func (f *factory) ProviderFor(ctx context.Context, pa v1alpha1.ProviderAccessor, c Config) (Provider, error) {
-	logger := log.FromContext(ctx)
-
-	var providerSecret *v1.Secret
-	if pa.IsDelegating() {
-		providerSecret = &v1.Secret{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      pa.GetProviderRef().Name,
-				Namespace: pa.GetNamespace(),
-			},
-			Type: v1alpha1.SecretTypeKuadrantEndpoint,
-			Data: map[string][]byte{
-				v1alpha1.EndpointLabelSelectorKey: []byte(fmt.Sprintf("%s=true, %s=%s", v1alpha1.AuthoritativeRecordLabel, v1alpha1.AuthoritativeRecordHashLabel, common.HashRootHost(pa.GetRootHost()))),
-			},
+// ProviderFor will return a Provider instance for the given ProviderAccessor(e.g. DNSRecord).
+// If the resource is delegating an in memory endpoint provider secret is returned with configuration appropriate for delegation of that resource.
+// If the resource is not delegating, and the secret does not exist in the resource namespace, an error will be returned.
+func (f *factory) ProviderFor(ctx context.Context, pa v1alpha1.ProviderAccessor, c Config) (provider Provider, err error) {
+	if !pa.IsDelegating() {
+		if pa.GetProviderRef().Name == "" {
+			return nil, errProviderRefRequired
 		}
-	} else {
-		providerSecret = &v1.Secret{
+		pSecret := &v1.Secret{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      pa.GetProviderRef().Name,
 				Namespace: pa.GetNamespace(),
 			}}
+		if err = f.Client.Get(ctx, client.ObjectKeyFromObject(pSecret), pSecret); err != nil {
+			return
+		}
 
-		if err := f.Client.Get(ctx, client.ObjectKeyFromObject(providerSecret), providerSecret); err != nil {
-			return nil, err
+		provider, err = f.providerForSecret(ctx, pSecret, c)
+		if err != nil {
```

Comment:
Slightly diff here with named return values. Took me a while to understand that was happening